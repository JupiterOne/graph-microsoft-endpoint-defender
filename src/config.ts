import {
  IntegrationExecutionContext,
  IntegrationInstanceConfig,
  IntegrationStepExecutionContext,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';
import { GraphClient } from './ms-graph/client';
import { IntegrationInstanceConfigFieldMap } from '@jupiterone/integration-sdk-core';

/**
 * A type describing the configuration fields required to execute the
 * integration for a specific account in the data provider.
 *
 * When executing the integration in a development environment, these values may
 * be provided in a `.env` file with environment variables. For example:
 *
 * - `CLIENT_ID=123` becomes `instance.config.clientId = '123'`
 * - `CLIENT_SECRET=abc` becomes `instance.config.clientSecret = 'abc'`
 *
 * Environment variables are NOT used when the integration is executing in a
 * managed environment. For example, in JupiterOne, users configure
 * `instance.config` in a UI.
 */

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  clientId: {
    type: 'string',
  },
  clientSecret: {
    type: 'string',
    mask: true,
  },
  tenant: {
    type: 'string',
  },
  findingSeverity: {
    type: 'string',
  },
  findingsLimit: {
    type: 'string',
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The Azure application client ID used to identify the program as the
   * registered app. This will be the same for all tenants since this is a
   * multi-tenant application.
   */
  clientId: string;

  /**
   * The Azuser App Registration client secret used to authenticate requests.
   */
  clientSecret: string;

  /**
   * The target directory/tenant ID, identified during the admin consent OAuth
   * flow.
   */
  tenant: string;
}

/**
 * An `IntegrationStepExecutionContext` typed for this integration's
 * `IntegrationInstanceConfig`.
 */
export type IntegrationStepContext =
  IntegrationStepExecutionContext<IntegrationConfig>;

export function validateExecutionConfig(
  executionContext: IntegrationExecutionContext<IntegrationConfig>,
): void {
  const { clientId, clientSecret, tenant } = executionContext.instance.config;

  executionContext.logger.info(
    {
      clientId,
      tenantId: tenant,
    },
    'Configured to make Microsoft Graph API calls to tenantId acting as clientId',
  );

  if (!clientId || !clientSecret || !tenant) {
    throw new IntegrationValidationError(
      'Config requires all of {clientId, clientSecret, tenant}',
    );
  }
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  validateExecutionConfig(context);
  const apiClient = new GraphClient(context.logger, context.instance.config);
  await apiClient.verifyAuthentication();
}
