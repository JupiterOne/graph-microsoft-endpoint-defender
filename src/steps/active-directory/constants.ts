import {
  RelationshipClass,
  RelationshipDirection,
} from '@jupiterone/integration-sdk-core';

export const steps: Record<string, string> = {
  FETCH_ACCOUNT: 'Fetch account',
  FETCH_MACHINE: 'Fetch machine',
  MACHINE_DEVICE_RELATIONSHIP: 'Map machine to device',
  FETCH_USERS: 'Fetch users',
  FETCH_FINDINGS: 'Fetch findings',
  FINDING_VULNERABILITY_RELATIONSHIP: 'Map finding to vulnerability',
};

export const entities = {
  ACCOUNT: {
    resourceName: 'Account',
    _type: 'microsoft_defender_account',
    _class: ['Account'],
  },
  MACHINE: {
    resourceName: 'Machine',
    _type: 'microsoft_defender_machine',
    _class: ['HostAgent'],
  },

  USER: {
    resourceName: 'User',
    _type: 'microsoft_defender_logon_user',
    _class: ['User'],
  },
  FINDING: {
    resourceName: 'Vulnerability',
    _type: 'microsoft_defender_vulnerability',
    _class: ['Finding'],
  },
};

export const TargetEntities = {
  CVE: {
    resourceName: 'CVE',
    _type: 'cve',
    _class: ['Vulnerability'],
  },
  DEVICE: {
    resourceName: 'Device',
    _type: 'user_endPoint',
    _class: ['Device/Host'],
  },
};

export const relationships = {
  ACCOUNT_HAS_MACHINE: {
    _type: 'microsoft_defender_account_has_machine',
    sourceType: entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: entities.MACHINE._type,
  },
  MACHINE_HAS_USER: {
    _type: 'microsoft_defender_machine_has_logon_user',
    sourceType: entities.MACHINE._type,
    _class: RelationshipClass.HAS,
    targetType: entities.USER._type,
  },
  MACHINE_IDENTIFIED_FINDING: {
    _type: 'microsoft_defender_machine_identified_vulnerability',
    sourceType: entities.MACHINE._type,
    _class: RelationshipClass.IDENTIFIED,
    targetType: entities.FINDING._type,
  },
};

export const MappedRelationships = {
  FINDING_IS_CVE_VULNERABILITY: {
    _type: 'microsoft_defender_vulnerability_is_cve',
    sourceType: entities.FINDING._type,
    _class: RelationshipClass.IS,
    targetType: TargetEntities.CVE._type,
    direction: RelationshipDirection.FORWARD,
  },
  MACHINE_MANAGES_DEVICE: {
    _type: 'microsoft_defender_machine_manages_user_endPoint',
    sourceType: entities.MACHINE._type,
    _class: RelationshipClass.MANAGES,
    targetType: TargetEntities.DEVICE._type,
    direction: RelationshipDirection.FORWARD,
  },
};

export const DATA_ACCOUNT_ENTITY = entities.ACCOUNT._type;
export const DATA_MACHINE_ENTITY = entities.MACHINE._type;
export const DATA_FINDING_ENTITY = entities.FINDING._type;
