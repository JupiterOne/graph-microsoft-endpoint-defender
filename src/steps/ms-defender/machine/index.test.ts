import { executeStepWithDependencies } from '@jupiterone/integration-sdk-testing';
import { Steps } from '../../../constants';
import { Recording, setupProjectRecording } from '../../../../test/recording';
import { buildStepTestConfigForStep } from '../../../../test/config';

let recording: Recording;
afterEach(async () => {
  await recording.stop();
});

test('#fetchMachines', async () => {
  recording = setupProjectRecording({
    directory: __dirname,
    name: '#fetchMachines',
  });

  const stepConfig = buildStepTestConfigForStep(Steps.FETCH_MACHINES.id);
  const stepResult = await executeStepWithDependencies(stepConfig);
  expect(stepResult).toMatchStepMetadata(stepConfig);
  expect(stepResult).toMatchSnapshot();
});

test('#fetchEndpoints', async () => {
  recording = setupProjectRecording({
    directory: __dirname,
    name: '#fetchEndpoints',
  });

  const stepConfig = buildStepTestConfigForStep(Steps.FETCH_ENDPOINTS.id);
  const stepResult = await executeStepWithDependencies(stepConfig);
  expect(stepResult).toMatchStepMetadata(stepConfig);
  expect(stepResult).toMatchSnapshot();
});
