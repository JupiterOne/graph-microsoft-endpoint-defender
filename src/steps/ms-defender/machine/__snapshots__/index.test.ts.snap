// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchEndpoints 1`] = `
Object {
  "collectedData": Object {},
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Device",
      ],
      "_key": "user_endpoint.595967ce83eb04bfddedaa7f29c8d10142a5bd30",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "@odata.context": "https://api.securitycenter.microsoft.com/api/$metadata#Machines/$entity",
            "aadDeviceId": null,
            "agentVersion": "10.8210.22000.978",
            "computerDnsName": "testmachine3",
            "defenderAvStatus": "NotSupported",
            "deviceValue": "Normal",
            "exposureLevel": "None",
            "firstSeen": "2022-09-13T10:46:16.911Z",
            "healthStatus": "Inactive",
            "id": "595967ce83eb04bfddedaa7f29c8d10142a5bd30",
            "ipAddresses": Array [
              Object {
                "ipAddress": "10.1.1.196",
                "macAddress": "000D3ADD1C52",
                "operationalStatus": "Up",
                "type": "Ethernet",
              },
              Object {
                "ipAddress": "fe80::b5b8:8c4a:400b:6144",
                "macAddress": "000D3ADD1C52",
                "operationalStatus": "Up",
                "type": "Ethernet",
              },
              Object {
                "ipAddress": "127.0.0.1",
                "macAddress": "",
                "operationalStatus": "Up",
                "type": "SoftwareLoopback",
              },
              Object {
                "ipAddress": "::1",
                "macAddress": "",
                "operationalStatus": "Up",
                "type": "SoftwareLoopback",
              },
            ],
            "isAadJoined": false,
            "lastExternalIpAddress": "20.12.71.218",
            "lastIpAddress": "10.1.1.196",
            "lastSeen": "2022-09-15T16:34:14.1307367Z",
            "machineTags": Array [
              "evaluation",
            ],
            "managedBy": "Unknown",
            "managedByStatus": "Unknown",
            "onboardingStatus": "Onboarded",
            "osArchitecture": "64-bit",
            "osBuild": 22000,
            "osPlatform": "Windows11",
            "osProcessor": "x64",
            "osVersion": null,
            "rbacGroupId": 0,
            "rbacGroupName": null,
            "riskScore": "None",
            "version": "21H2",
            "vmMetadata": Object {
              "cloudProvider": "Azure",
              "resourceId": "/subscriptions/1b922e46-8595-4749-997a-1205e714cd91/resourceGroups/evaluation_5547f014f83d461e95e529fe36fd0219_3/providers/Microsoft.Compute/virtualMachines/TestMachine3",
              "subscriptionId": "1b922e46-8595-4749-997a-1205e714cd91",
              "vmId": "fd7b83e4-ddec-4c0d-8026-418c8a3b9da8",
            },
          },
        },
      ],
      "_type": "user_endpoint",
      "aadDeviceId": null,
      "agentVersion": "10.8210.22000.978",
      "category": "endpoint",
      "cloudProvider": "Azure",
      "computerDnsName": "testmachine3",
      "createdOn": undefined,
      "defenderAvStatus": "NotSupported",
      "deviceId": "unknown",
      "deviceValue": "Normal",
      "displayName": "testmachine3",
      "exposureLevel": "None",
      "firstSeenOn": 1663065976911,
      "healthStatus": "Inactive",
      "id": "595967ce83eb04bfddedaa7f29c8d10142a5bd30",
      "ipAddresses": Array [
        "10.1.1.196",
        "fe80::b5b8:8c4a:400b:6144",
        "127.0.0.1",
        "::1",
      ],
      "isAadJoined": false,
      "lastExternalIpAddress": "20.12.71.218",
      "lastIpAddress": "10.1.1.196",
      "lastSeenOn": 1663259654130,
      "machineTags": Array [
        "evaluation",
      ],
      "make": "unknown",
      "managedBy": "Unknown",
      "managedByStatus": "Unknown",
      "model": "unknown",
      "name": "testmachine3",
      "onboardingStatus": "Onboarded",
      "osArchitecture": "64-bit",
      "osBuild": 22000,
      "osPlatform": "Windows11",
      "osProcessor": "x64",
      "osVersion": undefined,
      "rbacGroupId": 0,
      "rbacGroupName": null,
      "resourceId": "/subscriptions/1b922e46-8595-4749-997a-1205e714cd91/resourceGroups/evaluation_5547f014f83d461e95e529fe36fd0219_3/providers/Microsoft.Compute/virtualMachines/TestMachine3",
      "riskScore": "None",
      "serial": "unknown",
      "subscriptionId": "1b922e46-8595-4749-997a-1205e714cd91",
      "version": "21H2",
      "vmId": "fd7b83e4-ddec-4c0d-8026-418c8a3b9da8",
    },
    Object {
      "_class": Array [
        "Device",
      ],
      "_key": "user_endpoint.614f81ab67349d2e017f7f8bee5764e1f4a6397d",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "@odata.context": "https://api.securitycenter.microsoft.com/api/$metadata#Machines/$entity",
            "aadDeviceId": null,
            "agentVersion": "30.122071.17543.0",
            "computerDnsName": "testserver2",
            "defenderAvStatus": "NotSupported",
            "deviceValue": "Normal",
            "exposureLevel": "None",
            "firstSeen": "2022-09-13T08:57:51.994Z",
            "healthStatus": "Inactive",
            "id": "614f81ab67349d2e017f7f8bee5764e1f4a6397d",
            "ipAddresses": Array [
              Object {
                "ipAddress": "10.1.1.132",
                "macAddress": "6045BD8016FF",
                "operationalStatus": "Up",
                "type": "Other",
              },
              Object {
                "ipAddress": "fe80::6245:bdff:fe80:16ff",
                "macAddress": "6045BD8016FF",
                "operationalStatus": "Up",
                "type": "Other",
              },
              Object {
                "ipAddress": "127.0.0.1",
                "macAddress": "000000000000",
                "operationalStatus": "Up",
                "type": "Other",
              },
              Object {
                "ipAddress": "::1",
                "macAddress": "000000000000",
                "operationalStatus": "Up",
                "type": "Other",
              },
            ],
            "isAadJoined": false,
            "lastExternalIpAddress": "20.122.85.11",
            "lastIpAddress": "10.1.1.132",
            "lastSeen": "2022-09-16T08:58:29.4252411Z",
            "machineTags": Array [
              "evaluation",
            ],
            "managedBy": "Unknown",
            "managedByStatus": "Unknown",
            "onboardingStatus": "Onboarded",
            "osArchitecture": "64-bit",
            "osBuild": null,
            "osPlatform": "Ubuntu",
            "osProcessor": "x64",
            "osVersion": null,
            "rbacGroupId": 0,
            "rbacGroupName": null,
            "riskScore": "None",
            "version": "18.4",
            "vmMetadata": null,
          },
        },
      ],
      "_type": "user_endpoint",
      "aadDeviceId": null,
      "agentVersion": "30.122071.17543.0",
      "category": "endpoint",
      "cloudProvider": undefined,
      "computerDnsName": "testserver2",
      "createdOn": undefined,
      "defenderAvStatus": "NotSupported",
      "deviceId": "unknown",
      "deviceValue": "Normal",
      "displayName": "testserver2",
      "exposureLevel": "None",
      "firstSeenOn": 1663059471994,
      "healthStatus": "Inactive",
      "id": "614f81ab67349d2e017f7f8bee5764e1f4a6397d",
      "ipAddresses": Array [
        "10.1.1.132",
        "fe80::6245:bdff:fe80:16ff",
        "127.0.0.1",
        "::1",
      ],
      "isAadJoined": false,
      "lastExternalIpAddress": "20.122.85.11",
      "lastIpAddress": "10.1.1.132",
      "lastSeenOn": 1663318709425,
      "machineTags": Array [
        "evaluation",
      ],
      "make": "unknown",
      "managedBy": "Unknown",
      "managedByStatus": "Unknown",
      "model": "unknown",
      "name": "testserver2",
      "onboardingStatus": "Onboarded",
      "osArchitecture": "64-bit",
      "osBuild": null,
      "osPlatform": "Ubuntu",
      "osProcessor": "x64",
      "osVersion": undefined,
      "rbacGroupId": 0,
      "rbacGroupName": null,
      "resourceId": undefined,
      "riskScore": "None",
      "serial": "unknown",
      "subscriptionId": undefined,
      "version": "18.4",
      "vmId": undefined,
    },
    Object {
      "_class": Array [
        "Device",
      ],
      "_key": "user_endpoint.f39177bfecabebc12576671381d1f95c68492ae3",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "@odata.context": "https://api.securitycenter.microsoft.com/api/$metadata#Machines/$entity",
            "aadDeviceId": null,
            "agentVersion": "10.8040.19041.1889",
            "computerDnsName": "testmachine1",
            "defenderAvStatus": "Unknown",
            "deviceValue": "Normal",
            "exposureLevel": "None",
            "firstSeen": "2022-09-13T08:42:48.607Z",
            "healthStatus": "Inactive",
            "id": "f39177bfecabebc12576671381d1f95c68492ae3",
            "ipAddresses": Array [
              Object {
                "ipAddress": "10.1.1.68",
                "macAddress": "6045BD78564A",
                "operationalStatus": "Up",
                "type": "Ethernet",
              },
              Object {
                "ipAddress": "fe80::addd:6e16:244c:acc3",
                "macAddress": "6045BD78564A",
                "operationalStatus": "Up",
                "type": "Ethernet",
              },
              Object {
                "ipAddress": "127.0.0.1",
                "macAddress": "",
                "operationalStatus": "Up",
                "type": "SoftwareLoopback",
              },
              Object {
                "ipAddress": "::1",
                "macAddress": "",
                "operationalStatus": "Up",
                "type": "SoftwareLoopback",
              },
            ],
            "isAadJoined": false,
            "lastExternalIpAddress": "10.14.19.12",
            "lastIpAddress": "10.1.1.68",
            "lastSeen": "2022-09-13T08:53:41.095Z",
            "machineTags": Array [
              "evaluation",
            ],
            "managedBy": "Unknown",
            "managedByStatus": "Unknown",
            "onboardingStatus": "Onboarded",
            "osArchitecture": "64-bit",
            "osBuild": 19043,
            "osPlatform": "Windows10",
            "osProcessor": "x64",
            "osVersion": null,
            "rbacGroupId": 0,
            "rbacGroupName": null,
            "riskScore": "High",
            "version": "21H1",
            "vmMetadata": Object {
              "cloudProvider": "Azure",
              "resourceId": "/subscriptions/1b922e46-8595-4749-997a-1205e714cd91/resourceGroups/evaluation_5547f014f83d461e95e529fe36fd0219_1/providers/Microsoft.Compute/virtualMachines/TestMachine1",
              "subscriptionId": "1b922e46-8595-4749-997a-1205e714cd91",
              "vmId": "5841813e-5f75-4789-bd48-6cdee0592115",
            },
          },
        },
      ],
      "_type": "user_endpoint",
      "aadDeviceId": null,
      "agentVersion": "10.8040.19041.1889",
      "category": "endpoint",
      "cloudProvider": "Azure",
      "computerDnsName": "testmachine1",
      "createdOn": undefined,
      "defenderAvStatus": "Unknown",
      "deviceId": "unknown",
      "deviceValue": "Normal",
      "displayName": "testmachine1",
      "exposureLevel": "None",
      "firstSeenOn": 1663058568607,
      "healthStatus": "Inactive",
      "id": "f39177bfecabebc12576671381d1f95c68492ae3",
      "ipAddresses": Array [
        "10.1.1.68",
        "fe80::addd:6e16:244c:acc3",
        "127.0.0.1",
        "::1",
      ],
      "isAadJoined": false,
      "lastExternalIpAddress": "10.14.19.12",
      "lastIpAddress": "10.1.1.68",
      "lastSeenOn": 1663059221095,
      "machineTags": Array [
        "evaluation",
      ],
      "make": "unknown",
      "managedBy": "Unknown",
      "managedByStatus": "Unknown",
      "model": "unknown",
      "name": "testmachine1",
      "onboardingStatus": "Onboarded",
      "osArchitecture": "64-bit",
      "osBuild": 19043,
      "osPlatform": "Windows10",
      "osProcessor": "x64",
      "osVersion": undefined,
      "rbacGroupId": 0,
      "rbacGroupName": null,
      "resourceId": "/subscriptions/1b922e46-8595-4749-997a-1205e714cd91/resourceGroups/evaluation_5547f014f83d461e95e529fe36fd0219_1/providers/Microsoft.Compute/virtualMachines/TestMachine1",
      "riskScore": "High",
      "serial": "unknown",
      "subscriptionId": "1b922e46-8595-4749-997a-1205e714cd91",
      "version": "21H1",
      "vmId": "5841813e-5f75-4789-bd48-6cdee0592115",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "microsoft_defender_machine:595967ce83eb04bfddedaa7f29c8d10142a5bd30",
      "_key": "microsoft_defender_machine:595967ce83eb04bfddedaa7f29c8d10142a5bd30|manages|user_endpoint.595967ce83eb04bfddedaa7f29c8d10142a5bd30",
      "_toEntityKey": "user_endpoint.595967ce83eb04bfddedaa7f29c8d10142a5bd30",
      "_type": "microsoft_defender_machine_manages_user_endpoint",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "microsoft_defender_machine:614f81ab67349d2e017f7f8bee5764e1f4a6397d",
      "_key": "microsoft_defender_machine:614f81ab67349d2e017f7f8bee5764e1f4a6397d|manages|user_endpoint.614f81ab67349d2e017f7f8bee5764e1f4a6397d",
      "_toEntityKey": "user_endpoint.614f81ab67349d2e017f7f8bee5764e1f4a6397d",
      "_type": "microsoft_defender_machine_manages_user_endpoint",
      "displayName": "MANAGES",
    },
    Object {
      "_class": "MANAGES",
      "_fromEntityKey": "microsoft_defender_machine:f39177bfecabebc12576671381d1f95c68492ae3",
      "_key": "microsoft_defender_machine:f39177bfecabebc12576671381d1f95c68492ae3|manages|user_endpoint.f39177bfecabebc12576671381d1f95c68492ae3",
      "_toEntityKey": "user_endpoint.f39177bfecabebc12576671381d1f95c68492ae3",
      "_type": "microsoft_defender_machine_manages_user_endpoint",
      "displayName": "MANAGES",
    },
  ],
  "encounteredEntityKeys": Set {
    "user_endpoint.595967ce83eb04bfddedaa7f29c8d10142a5bd30",
    "user_endpoint.614f81ab67349d2e017f7f8bee5764e1f4a6397d",
    "user_endpoint.f39177bfecabebc12576671381d1f95c68492ae3",
    "microsoft_defender_account:local-integration-instance",
    "microsoft_defender_machine:595967ce83eb04bfddedaa7f29c8d10142a5bd30",
    "microsoft_defender_machine:614f81ab67349d2e017f7f8bee5764e1f4a6397d",
    "microsoft_defender_machine:f39177bfecabebc12576671381d1f95c68492ae3",
  },
  "encounteredTypes": Array [
    "microsoft_defender_account",
    "microsoft_defender_machine",
    "microsoft_defender_account_has_machine",
    "user_endpoint",
    "microsoft_defender_machine_manages_user_endpoint",
  ],
}
`;

exports[`#fetchMachines 1`] = `
Object {
  "collectedData": Object {},
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "HostAgent",
      ],
      "_key": "microsoft_defender_machine:595967ce83eb04bfddedaa7f29c8d10142a5bd30",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "aadDeviceId": null,
            "agentVersion": "10.8040.22000.832",
            "computerDnsName": "testmachine3",
            "defenderAvStatus": "NotSupported",
            "deviceValue": "Normal",
            "exposureLevel": "Medium",
            "firstSeen": "2022-09-13T10:46:16.911Z",
            "healthStatus": "Active",
            "id": "595967ce83eb04bfddedaa7f29c8d10142a5bd30",
            "ipAddresses": Array [
              Object {
                "ipAddress": "10.1.1.196",
                "macAddress": "000D3ADD1C52",
                "operationalStatus": "Up",
                "type": "Ethernet",
              },
              Object {
                "ipAddress": "fe80::b5b8:8c4a:400b:6144",
                "macAddress": "000D3ADD1C52",
                "operationalStatus": "Up",
                "type": "Ethernet",
              },
              Object {
                "ipAddress": "127.0.0.1",
                "macAddress": "",
                "operationalStatus": "Up",
                "type": "SoftwareLoopback",
              },
              Object {
                "ipAddress": "::1",
                "macAddress": "",
                "operationalStatus": "Up",
                "type": "SoftwareLoopback",
              },
            ],
            "isAadJoined": false,
            "lastExternalIpAddress": "20.12.71.218",
            "lastIpAddress": "10.1.1.196",
            "lastSeen": "2022-09-13T20:05:43.9868473Z",
            "machineTags": Array [
              "evaluation",
            ],
            "managedBy": "Unknown",
            "managedByStatus": "Unknown",
            "onboardingStatus": "Onboarded",
            "osArchitecture": "64-bit",
            "osBuild": 22000,
            "osPlatform": "Windows11",
            "osProcessor": "x64",
            "osVersion": null,
            "rbacGroupId": 0,
            "rbacGroupName": null,
            "riskScore": "None",
            "version": "21H2",
            "vmMetadata": Object {
              "cloudProvider": "Azure",
              "resourceId": "/subscriptions/1b922e46-8595-4749-997a-1205e714cd91/resourceGroups/evaluation_5547f014f83d461e95e529fe36fd0219_3/providers/Microsoft.Compute/virtualMachines/TestMachine3",
              "subscriptionId": "1b922e46-8595-4749-997a-1205e714cd91",
              "vmId": "fd7b83e4-ddec-4c0d-8026-418c8a3b9da8",
            },
          },
        },
      ],
      "_type": "microsoft_defender_machine",
      "agentVersion": "10.8040.22000.832",
      "computerDnsName": "testmachine3",
      "createdOn": undefined,
      "defenderAvStatus": "NotSupported",
      "displayName": "testmachine3",
      "function": Array [
        "endpoint-compliance",
        "endpoint-configuration",
        "endpoint-protection",
        "anti-malware",
        "vulnerability-detection",
        "container-security",
      ],
      "id": "595967ce83eb04bfddedaa7f29c8d10142a5bd30",
      "machineTags": Array [
        "evaluation",
      ],
      "managedBy": "Unknown",
      "managedByStatus": "Unknown",
      "name": "testmachine3",
      "onboardingStatus": "Onboarded",
      "rbacGroupId": 0,
      "rbacGroupName": null,
      "riskScore": "None",
    },
    Object {
      "_class": Array [
        "HostAgent",
      ],
      "_key": "microsoft_defender_machine:614f81ab67349d2e017f7f8bee5764e1f4a6397d",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "aadDeviceId": null,
            "agentVersion": "30.122071.17543.0",
            "computerDnsName": "testserver2",
            "defenderAvStatus": "NotSupported",
            "deviceValue": "Normal",
            "exposureLevel": "Medium",
            "firstSeen": "2022-09-13T08:57:51.994Z",
            "healthStatus": "Active",
            "id": "614f81ab67349d2e017f7f8bee5764e1f4a6397d",
            "ipAddresses": Array [
              Object {
                "ipAddress": "10.1.1.132",
                "macAddress": "6045BD8016FF",
                "operationalStatus": "Up",
                "type": "Other",
              },
              Object {
                "ipAddress": "fe80::6245:bdff:fe80:16ff",
                "macAddress": "6045BD8016FF",
                "operationalStatus": "Up",
                "type": "Other",
              },
              Object {
                "ipAddress": "127.0.0.1",
                "macAddress": "000000000000",
                "operationalStatus": "Up",
                "type": "Other",
              },
              Object {
                "ipAddress": "::1",
                "macAddress": "000000000000",
                "operationalStatus": "Up",
                "type": "Other",
              },
            ],
            "isAadJoined": false,
            "lastExternalIpAddress": "20.122.85.11",
            "lastIpAddress": "10.1.1.132",
            "lastSeen": "2022-09-14T00:58:16.964238Z",
            "machineTags": Array [
              "evaluation",
            ],
            "managedBy": "Unknown",
            "managedByStatus": "Unknown",
            "onboardingStatus": "Onboarded",
            "osArchitecture": "64-bit",
            "osBuild": null,
            "osPlatform": "Ubuntu",
            "osProcessor": "x64",
            "osVersion": null,
            "rbacGroupId": 0,
            "rbacGroupName": null,
            "riskScore": "None",
            "version": "18.4",
            "vmMetadata": null,
          },
        },
      ],
      "_type": "microsoft_defender_machine",
      "agentVersion": "30.122071.17543.0",
      "computerDnsName": "testserver2",
      "createdOn": undefined,
      "defenderAvStatus": "NotSupported",
      "displayName": "testserver2",
      "function": Array [
        "endpoint-compliance",
        "endpoint-configuration",
        "endpoint-protection",
        "anti-malware",
        "vulnerability-detection",
        "container-security",
      ],
      "id": "614f81ab67349d2e017f7f8bee5764e1f4a6397d",
      "machineTags": Array [
        "evaluation",
      ],
      "managedBy": "Unknown",
      "managedByStatus": "Unknown",
      "name": "testserver2",
      "onboardingStatus": "Onboarded",
      "rbacGroupId": 0,
      "rbacGroupName": null,
      "riskScore": "None",
    },
    Object {
      "_class": Array [
        "HostAgent",
      ],
      "_key": "microsoft_defender_machine:f39177bfecabebc12576671381d1f95c68492ae3",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "aadDeviceId": null,
            "agentVersion": "10.8040.19041.1889",
            "computerDnsName": "testmachine1",
            "defenderAvStatus": "Updated",
            "deviceValue": "Normal",
            "exposureLevel": "Medium",
            "firstSeen": "2022-09-13T08:42:48.607Z",
            "healthStatus": "Active",
            "id": "f39177bfecabebc12576671381d1f95c68492ae3",
            "ipAddresses": Array [
              Object {
                "ipAddress": "10.1.1.68",
                "macAddress": "6045BD78564A",
                "operationalStatus": "Up",
                "type": "Ethernet",
              },
              Object {
                "ipAddress": "fe80::addd:6e16:244c:acc3",
                "macAddress": "6045BD78564A",
                "operationalStatus": "Up",
                "type": "Ethernet",
              },
              Object {
                "ipAddress": "127.0.0.1",
                "macAddress": "",
                "operationalStatus": "Up",
                "type": "SoftwareLoopback",
              },
              Object {
                "ipAddress": "::1",
                "macAddress": "",
                "operationalStatus": "Up",
                "type": "SoftwareLoopback",
              },
            ],
            "isAadJoined": false,
            "lastExternalIpAddress": "10.14.19.12",
            "lastIpAddress": "10.1.1.68",
            "lastSeen": "2022-09-13T08:53:41.095Z",
            "machineTags": Array [
              "evaluation",
            ],
            "managedBy": "Unknown",
            "managedByStatus": "Unknown",
            "onboardingStatus": "Onboarded",
            "osArchitecture": "64-bit",
            "osBuild": 19043,
            "osPlatform": "Windows10",
            "osProcessor": "x64",
            "osVersion": null,
            "rbacGroupId": 0,
            "rbacGroupName": null,
            "riskScore": "High",
            "version": "21H1",
            "vmMetadata": Object {
              "cloudProvider": "Azure",
              "resourceId": "/subscriptions/1b922e46-8595-4749-997a-1205e714cd91/resourceGroups/evaluation_5547f014f83d461e95e529fe36fd0219_1/providers/Microsoft.Compute/virtualMachines/TestMachine1",
              "subscriptionId": "1b922e46-8595-4749-997a-1205e714cd91",
              "vmId": "5841813e-5f75-4789-bd48-6cdee0592115",
            },
          },
        },
      ],
      "_type": "microsoft_defender_machine",
      "agentVersion": "10.8040.19041.1889",
      "computerDnsName": "testmachine1",
      "createdOn": undefined,
      "defenderAvStatus": "Updated",
      "displayName": "testmachine1",
      "function": Array [
        "endpoint-compliance",
        "endpoint-configuration",
        "endpoint-protection",
        "anti-malware",
        "vulnerability-detection",
        "container-security",
      ],
      "id": "f39177bfecabebc12576671381d1f95c68492ae3",
      "machineTags": Array [
        "evaluation",
      ],
      "managedBy": "Unknown",
      "managedByStatus": "Unknown",
      "name": "testmachine1",
      "onboardingStatus": "Onboarded",
      "rbacGroupId": 0,
      "rbacGroupName": null,
      "riskScore": "High",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "microsoft_defender_account:local-integration-instance",
      "_key": "microsoft_defender_account:local-integration-instance|has|microsoft_defender_machine:595967ce83eb04bfddedaa7f29c8d10142a5bd30",
      "_toEntityKey": "microsoft_defender_machine:595967ce83eb04bfddedaa7f29c8d10142a5bd30",
      "_type": "microsoft_defender_account_has_machine",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "microsoft_defender_account:local-integration-instance",
      "_key": "microsoft_defender_account:local-integration-instance|has|microsoft_defender_machine:614f81ab67349d2e017f7f8bee5764e1f4a6397d",
      "_toEntityKey": "microsoft_defender_machine:614f81ab67349d2e017f7f8bee5764e1f4a6397d",
      "_type": "microsoft_defender_account_has_machine",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "microsoft_defender_account:local-integration-instance",
      "_key": "microsoft_defender_account:local-integration-instance|has|microsoft_defender_machine:f39177bfecabebc12576671381d1f95c68492ae3",
      "_toEntityKey": "microsoft_defender_machine:f39177bfecabebc12576671381d1f95c68492ae3",
      "_type": "microsoft_defender_account_has_machine",
      "displayName": "HAS",
    },
  ],
  "encounteredEntityKeys": Set {
    "microsoft_defender_machine:595967ce83eb04bfddedaa7f29c8d10142a5bd30",
    "microsoft_defender_machine:614f81ab67349d2e017f7f8bee5764e1f4a6397d",
    "microsoft_defender_machine:f39177bfecabebc12576671381d1f95c68492ae3",
    "microsoft_defender_account:local-integration-instance",
  },
  "encounteredTypes": Array [
    "microsoft_defender_account",
    "microsoft_defender_machine",
    "microsoft_defender_account_has_machine",
  ],
}
`;
